### ISSUE TITLE
Getting next token

### ISSUE THREAD

**Issue by farid-fari** [andrewpeng02/transformer-translation#4]
Hey, I've been trying to understand the way the transformer from pytorch works and using your work to better understand.

Specifically, how you use the output to get the next token:

https://github.com/andrewpeng02/transformer-translation/blob/d735f04bee6355d67c87f1ea3567de65f9291b46/translate-sentence.py#L27

Since in the model you reshape to `n t e`:

https://github.com/andrewpeng02/transformer-translation/blob/d735f04bee6355d67c87f1ea3567de65f9291b46/model.py#L46

It would seem that you're selecting the last in the batch (with the `[-1]`) and the first value from the target length dimension with the `[0]`.

So I guess my question is, why? Why does the first value from the output represent what should be the next token? I don't even understand how it is done in a usual transformer, since the last decoder should output `t` vectors of shape `e`, how do we feed them into a fixed linear layer?

It would make more sense to me to use the last one, but I still don't understand why we're discarding `t-1` of these vectors...

**Comment by andrewpeng02** [comment]
After rearranging on line 46, `output` is of shape (number of sentences, number of tokens, embedding dim). The fc layer acts on the last dimension, meaning the output of the fc layer is (number of sentences, number of tokens, tgt vocab size). In effect, the fc layer is used `number of sentences * number of tokens` per pass. 

In `translate-sentence.py`, I take the output of the fc layer, get the first sentence, and use `torch.topk` (https://pytorch.org/docs/stable/generated/torch.topk.html). `torch.topk(output, 5)` returns 2 tensors of shape (number of tokens, 5). So `indices[-1]` gets the last token in the output and then `[0]` gives the index with the largest value. I hope this helps!

**Comment by farid-fari** [comment]
Okay, I now understand how the `[0]` gets the top 1 value from the `topk`, however I'm still confused at which point you get the first sentence, i.e. get rid of the batch size dimension?

Since `output` has the batch dimension as first dimension, then `indices` should as well, and so `indices[-1][0]` should get the last sentence by batch and the first token in the sentence, right?

Besides this implementation question, do you know that this is how most transformers will work, by using the last token in the translated sequence to feed to the linear `fc` layer and softmax layer?

**Comment by andrewpeng02** [comment]
In the `forward_model(model, src, tgt)` method I return `output.squeeze(0).to('cpu')`. Because the first dimension is only 1, this removes the first dimension and I'm left with (number of tokens, tgt vocab size). In `translate-sentence.py`, I only pass 1 sentence to the model at a time.

Yes, when translating from scratch, you'll have to generate the translated sentences one token at a time. The last token represents the next token in the translated sentence. We already have all of the other tokens, that's what we passed in.

**Comment by farid-fari** [comment]
Oh okay, I had missed that return with the `squeeze`, that explains it for me. Thanks a lot!
